presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsAnnotationProcessor.java

```diff
@@ -0,0 +1,255 @@
+                statValue = NaN;
+            }
+        }
+        else {
+            switch (operation) {
+                case USE_SOURCE_STATS:
+                    statValue = sourceStats.get(sourceStatsArgumentIndex);
+                    break;
+                case ROW_COUNT:
+                    statValue = outputRowCount;
+                    break;
+                case NON_NULL_ROW_COUNT:
+                    statValue = outputRowCount * (1 - nullFraction);
+                    break;
+                case USE_TYPE_WIDTH_VARCHAR:
+                    statValue = typeWidthVarchar(callExpression, sourceStatsArgumentIndex);
+                    break;
+            }
+        }
+        return statValue;
+    }
+
+    private double typeWidthVarchar(CallExpression call, int argumentIndex)
+    {
+        TypeSignature typeSignature = call.getArguments().get(argumentIndex).getType().getTypeSignature();
+        if (typeSignature.getTypeSignatureBase().hasStandardType() && typeSignature.getTypeSignatureBase().getStandardTypeBase().equals(
+                StandardTypes.VARCHAR)) {
+            if (typeSignature.getParameters().size() == 1) { // Varchar type should have 1 parameter i.e. size.;
+                Long longLiteral = typeSignature.getParameters().get(0).getLongLiteral();
+                if (longLiteral > 0 && longLiteral != VarcharType.UNBOUNDED_LENGTH) {
+                    return longLiteral;
+                }
+            }
+        }
+        return NaN;
+    }
+
+    private double getReturnTypeWidth(CallExpression call, StatsPropagationBehavior operation)
+    {
+        if (call.getType() instanceof FixedWidthType) {
+            return ((FixedWidthType) call.getType()).getFixedSize();
+        }
+        if (call.getType() instanceof VarcharType) {
+            VarcharType returnType = (VarcharType) call.getType();
+            if (!returnType.isUnbounded()) {
+                return returnType.getLengthSafe();
+            }
+            if (operation == SUM_ARGUMENTS) {
+                // since return type is a varchar and length is unknown,
+                // getting it by doing a SUM each argument's varchar type bounds.
+                double sum = 0;
+                for (RowExpression r : call.getArguments()) {
+                    if (r instanceof CallExpression) { // argument is another function call
+                        sum += getReturnTypeWidth((CallExpression) r, UNKNOWN);
+                    }
+                    if (r.getType() instanceof VarcharType) {
+                        VarcharType argType = (VarcharType) r.getType();
+                        if (!argType.isUnbounded()) {
+                            sum += argType.getLengthSafe();
+                        }
+                    }
+                }
+                if (sum > 0) {
+                    return sum;
+                }
+            }
+        }
+        return NaN;
+    }
+
+    private double firstNonNaN(double... values)
```

Comment:
> * This can be moved to MoreMath class too. That class already has a `firstNonNaN` method. Will that suffice ?

No, because that one does not allow all NaNs in arguments.

>     * This function will return `NaN` if all values are `NaN`. Is this expected ?

Yes
> 
>     * You can return early when the first `NaN` is found instead of iterating through the remaining values

Yes